Container proxy

bqqbarbhg opened the pull request:
Add proxy for routing requests from subdomain to user app containers. Added also initial database schema for containers table. The constraint was added to a previous migration as SQLite can't handle adding constraints later to tables, we need to figure out what to do about this.

The default settings for the Go proxy work with the development Node server, ie. compatible with `cd server && npm start`. The proxy doesn't support Postgres yet, but it should literally be one line change in the imports of `resolver.go`.

The proxy uses VGo modules so you need at least Go 1.11 to build this. With a compatible Go version building and running should work as simply as:
```bash
cd proxy
go build
./proxy
```

Note that if you are using Windows you may have issues with building the sqlite3 dependency as it's a CGo package that requires `gcc` to be in path. Moreover you may need 64-bit `gcc` (at least I did). I got it working with Mingw64: https://sourceforge.net/projects/mingw-w64/

bqqbarbhg committed:
Add initial DB container resolving

bqqbarbhg committed:
Somewhat working Go proxy

bqqbarbhg committed:
Integrate proxy with development server

bqqbarbhg committed:
Add constraints

bqqbarbhg committed:
Rename containers.go -> resolver.go

bqqbarbhg committed:
Clean up proxy into modules

bqqbarbhg commented:
Oh damn tests broke due to adding the `FOREIGN KEY` constraint to the threads table, will fix tomorrow!

bqqbarbhg committed:
Merge branch 'develop' into proxy

bqqbarbhg committed:
Fix test FOREIGN KEYS

bqqbarbhg committed:
Mark *.sh files with LF ends

bqqbarbhg committed:
Re-order production API Dockerfile for better caching

bqqbarbhg committed:
Rename proxy to server

bqqbarbhg committed:
Add Postgres support for proxy

bqqbarbhg committed:
Add proxy to Docker

tobbe-j commented on proxy/resolver/resolver.go:
```diff
@@ -0,0 +1,219 @@
+
+// Query the database and create a container instance
+func fetchContainerFromDatabase(id string) (*Container, error) {
+	row := db.QueryRow("SELECT subdomain, url FROM containers WHERE subdomain = ?", id)
```
Voi tulla ongelmia Postgres:in kanssa, nodessa postgres käytti $1,$2... '?','?' sijaan

bqqbarbhg replied:
Will test when I can merge Postgres here

tobbe-j commented on server/src/database/migrations/002-containers.sql:
```diff
@@ -0,0 +1,16 @@
+--
+BEGIN TRANSACTION;
+
+-- Table: containers
```
Lisää myös database docsiin

bqqbarbhg replied:
Done

tobbe-j reviewed (changes requested):
Docs

bqqbarbhg committed:
Inject script before </body>

bqqbarbhg committed:
Add containers to docs

bqqbarbhg committed:
DB query dependent on driver

bqqbarbhg committed:
Merge branch 'feature-postgres' into proxy

bqqbarbhg committed:
Fix Proxy in Docker

- Fix Postgres SQL syntax
- Add environment substitution for $APP_URL

bqqbarbhg committed:
Change default APP_URL

bqqbarbhg committed:
Merge branch 'develop' into proxy

bqqbarbhg committed:
Add container proxy tests to Travis

bqqbarbhg committed:
Python 2 -> 3, yaml -> yml

bqqbarbhg committed:
Add local proxy test + docker hostnames

bqqbarbhg committed:
Change yaml -> yml

bqqbarbhg committed:
Manually download Go 1.11 + fix subdomains

bqqbarbhg committed:
Fix reference script injections

bqqbarbhg committed:
Add Go executable to path

bqqbarbhg committed:
Download go to build/ folder

bqqbarbhg committed:
Print Go version

bqqbarbhg committed:
Reset $GOROOT

bqqbarbhg committed:
Fix whitespace diff error

bqqbarbhg committed:
Fix proxy curl test directory

bqqbarbhg committed:
Add proxy to development docker

bqqbarbhg committed:
Remove debug ls

bqqbarbhg committed:
Fix development proxy

bqqbarbhg committed:
Fix Docker start command

bqqbarbhg committed:
Lint fixes, remove unused package

bqqbarbhg commented:
Now this PR should be in a working condition.

Changes:
- Added support for using the proxy in a development server
- Added `--startProxy` argument to server for starting the proxy executable as a child process
- Added support for using the proxy in the development Docker build
- Added Travis CI tests for the proxy

bqqbarbhg commented:
Also includes the following misc fixes:
- Only initialize the feedback tool once even if it's imported multiple times via `<script>`
- Retry Postgres migrations a few times with timeout in case the DB server is still booting

tobbe-j reviewed (approved):
Looks good

HenrikHelenius commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(25/89) [task] [dynamic docker] Make dynamic docker Samuli](https://trello.com/c/LN926OBv/129-25-89-task-dynamic-docker-make-dynamic-docker-samuli)
